lisa/tests/hotplug.py

```diff
@@ -269,4 +271,174 @@ def cpuhp_seq(cls, nr_operations, hotpluggable_cpus, max_cpus_off, random_gen):
+            cls._mark_trace(target, ftrace_coll, start=0)
+
+            # Test hotunplug rollback for each possible state failure
+            for failing_state in states_down:
+                cls._test_rollback(target, ftrace_coll, cpu, failing_state,
+                    up=False)
+
+            # Test hotplug rollback for each possible state failure
+            for failing_state in states_up:
+                cls._test_rollback(target, ftrace_coll, cpu, failing_state,
+                    up=True)
+
+        ftrace_coll.get_data(ArtifactPath.join(res_dir, cls.TRACE_PATH))
+
+        return cls(res_dir, target.plat_info)
+
+    @classmethod
+    def _get_expected_states(cls, df, up):
+        df = df[(df['expected'] == 1.0) & (df['up'] == str(up))]
+
+        return df['idx'].dropna().tolist()
+
+    @requires_events('userspace@HotplugRollback', 'cpuhp_enter')
+    def test_hotplug_rollback(self) -> ResultBundle:
+        """
+        Test that the hotplug can rollback to its previous state after a
+        failure. All possible steps, up/down combinations will be tested. For
+        each combination, also verify that the hotplug is going through all the
+        steps it is supposed to.
+        """
+        df = df_merge([
+            self.trace.df_event('userspace@HotplugRollback'),
+            self.trace.df_event('cpuhp_enter')
+        ])
+
+        # Keep only the states delimited by _mark_trace()
+        df['test'].ffill(inplace=True)
+        df = df[df['test'] == 1]
+
+        # Read the expected states from full hot(un)plug
+        df['expected'].ffill(inplace=True)
+        df['up'].ffill(inplace=True)
+        expected_down = self._get_expected_states(df, up=False)
+        expected_up = self._get_expected_states(df, up=True)
+
+        def _get_expected_rollback(up, failing_state):
+            if up == 'True':
+                expected = [
+                    s for s in expected_up if s < int(failing_state)]
+                expected.extend([
+                    s for s in expected_down if s < int(failing_state)
+                ])
+            else:
+                expected = [
+                    s for s in expected_down if s > int(failing_state)]
+                expected.extend([
+                    s for s in expected_up if s > int(failing_state)
+                ])
+
+            return expected
+
+        def _verify_rollback(df):
+            failing_state = df['failing_state'].iloc[0]
+            up = df['up'].iloc[0]
+            expected = _get_expected_rollback(up, failing_state)
+
+            return pd.DataFrame(data={
+                'failing_state': df['failing_state'],
+                'up': up,
+                'result': df['idx'].tolist() == expected
+            })
+
+        df['failing_state'].ffill(inplace=True)
+        df.dropna(inplace=True)
+        df = df[df['expected'] == 0.0]
+        df = df.groupby(['up', 'failing_state']).apply(_verify_rollback)
```

Comment:
should pass `observed=True` to all groupby() calls to be more robust against categorical dtypes